### PR TITLE
feat: add method to add metadata section values as a dictionary

### PR DESCRIPTION
**Pull request by robinmacharg**
## Goal

Adds a `BugsnagMetadata.addMetadataToSection:values:` method to allow merging arbitrary dictionaries with preexisting Metadata.

## Design

Simple fast enumeration loop over provided values with appropriate checks for:

- Validity of Key (String subclass)
- Validity of Value (can be sanitised: a recursive check that all (nested) values are of type Number, String, Array, Dictionary or Null

Conditional addition/deletion depending on whether value is null or not.  There is a check to ensure that empty sections are not created for (failing) invalid values.  Invalid values are logged and ignored.  Delegate callback is called regardless of whether the method actually made a material change to the metadata.

## Changeset

- Method added to `BugsnagMetadata`.
- Unit Tests added.

## Tests

- Added Unit tests only.  This change affects the framework API and doesn't change any external aspect.
- CI passes.

### Outstanding Questions

- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

- [ ] The correct target branch has been selected (`master` for fixes, `next` for
  features)
- [ ] Consistency across platforms for structures or concepts added or modified
- [X] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [X] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [X] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: Tests/BugsnagMetadataTest.m [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+@end
+
+// MARK: - Tests
+
+@implementation BugsnagMetadataTest
+
+@synthesize delegateCalled;
+
+- (void)test_addAttribute_withName {
```
Comment (fractalwrench) [inline Tests/BugsnagMetadataTest.m]: The contents of this test looks good - if it's straightforward would it be possible to separate these test cases into individual functions? E.g., one for creation, one for checking an arbitrary tab name creates itself, etc.

File: Tests/BugsnagMetadataTest.m [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    tab2 = [metadata getTab:@"FirstTab"];
+    XCTAssertEqual(tab2.count, 0);
+    XCTAssertTrue(delegateCalled, "Expected the delegate's metadataChanged: method to be called.");
+
+    // Adding invalid values should fail silently (and not add the value)
+    delegateCalled = NO;
+    [metadata addAttribute:@"bar" withValue:[DummyClass new] toTabWithName:@"FirstTab"];
+    tab2 = [metadata getTab:@"FirstTab"];
+    XCTAssertEqual(tab2.count, 0);
+    XCTAssertTrue(delegateCalled, "Expected the delegate's metadataChanged: method to be called.");
```
Comment (fractalwrench) [inline Tests/BugsnagMetadataTest.m]: If _all_ the values are invalid and nothing has changed in the metadata, then we could probably avoid calling the delegate here - particularly if it does a lot of work. I'm unsure if this is realistically going to happen often, so would be good to get your judgement on whether you think it's worth making that change.

File: Tests/BugsnagMetadataTest.m (hunk 1 above)
Comment (robinmacharg) [reply Tests/BugsnagMetadataTest.m]: NP, extended setUp() slightly, split tests up and adjusted any cumulative values being tested to account for this.

File: Tests/BugsnagMetadataTest.m (hunk 2 above)
Comment (robinmacharg) [reply Tests/BugsnagMetadataTest.m]: Agreed.  I'd noted that.  It's small so I've made the change, ensuring that any delegate calls occur at most once per operation, and outside @synchronised sections.  Two tests had to be adjusted to check that the delegate is NOT called.
